fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt

```diff
@@ -46,21 +50,82 @@ class ReviewQualityCheckNetworkMiddleware(
-                    store.dispatch(ReviewQualityCheckAction.UpdateProductReview(productReviewState))
+                    logger.info("Fetching product analysis")
+                    val productReviewState = fetchAnalysis()
+                    logger.debug("Product review state: $productReviewState")
+                    store.updateProductReviewState(productReviewState)
                 }
             }
 
             ReviewQualityCheckAction.ReanalyzeProduct -> {
-                // Bug 1853311 - Integrate analyze and analysis_status
+                scope.launch {
+                    val reanalysis = reviewQualityCheckService.reanalyzeProduct().also {
+                        logger.debug("Reanalysis: $it")
+                    }
+
+                    if (reanalysis == null) {
+                        store.updateProductReviewState(ProductReviewState.Error.GenericError)
+                        return@launch
+                    }
+
+                    val status = pollForAnalysisStatus()
+
+                    if (status == null ||
+                        status == AnalysisStatusDto.PENDING ||
+                        status == AnalysisStatusDto.IN_PROGRESS
+                    ) {
+                        // poll failed, reset to previous state
+                        val state = store.state
+                        if (state is ReviewQualityCheckState.OptedIn) {
+                            if (state.productReviewState is ProductReviewState.NoAnalysisPresent) {
+                                store.updateProductReviewState(ProductReviewState.NoAnalysisPresent())
+                            } else if (state.productReviewState is ProductReviewState.AnalysisPresent) {
+                                store.updateProductReviewState(
+                                    state.productReviewState.copy(
+                                        analysisStatus = AnalysisStatus.NEEDS_ANALYSIS,
+                                    ),
+                                )
+                            }
+                        }
+                        return@launch
+                    } else {
+                        // poll succeeded, update state
+                        store.updateProductReviewState(status.toProductReviewState())
+                    }
+                }
             }
         }
     }
+
+    private suspend fun fetchAnalysis(): ProductReviewState =
+        reviewQualityCheckService.fetchProductReview().toProductReviewState()
+
+    private suspend fun pollForAnalysisStatus(): AnalysisStatusDto? =
+        retry(
+            predicate = { it == AnalysisStatusDto.PENDING || it == AnalysisStatusDto.IN_PROGRESS },
+            func = { reviewQualityCheckService.analysisStatus() },
+        ).also {
+            logger.debug("Status: $it")
+        }
+
+    private suspend fun AnalysisStatusDto.toProductReviewState(): ProductReviewState =
+        when (this) {
+            AnalysisStatusDto.COMPLETED -> fetchAnalysis()
```

Comment:
can we maybe restructure this + `fetchAnalysis` slightly? they currently have a bit of a circular dependency 😨 

Maybe have the callers of `fetchAnalysis` call `toProductReviewState` instead, so the fetching is an isolated function and the transformation is a separate distinct step?

It feels off to have `toProductReviewState`, which feels like a state transformation/mapping function, firing off the network call, call itself, and return the result of calling itself back to `AnalysisStatusDto.COMPLETED`.

Reply:
They don't have a circular dependency – `fetchAnalysis` is calling `ProductAnalysis.toProductReviewState()` not this function. This function calls `fetchAnalysis`.

It's data logic that's acting on the response of one endpoint - deciding to fetch when the analysis is completed and returns errors in other cases. 

And yes, I think we agree that a mapper is a good place for functions that are just transformations, that's also another reason why this is not in `ProductAnalysisMapper`.

Reply:
Ahh gotcha gotcha, sorry for the confusion on my end then. This looks good 👍